### PR TITLE
First try at refactor of Prepare.targets

### PR DESCRIPTION
**Pull request by piotrpMSFT**
@jgoshi @livarcocc @jonsequitur @eerhardt @krwq 

With @jgoshi's #5617 merged, I took a first pass at refactoring Prepare.targets into statically evaluated props files. Here are some things I'm experimenting with:

- DO NOT create a Prepare.props. Instead, decompose the props into files that represent responsibilities. I think that Version.props is a good example. Stage0.props probably needs some naming, if not intention, changes as it intends to cover "the stuff we need to build this repo"
- DO make the root `dir.props` the aggregator of these props. Ordering for pulling in the props files is relevant and that seems like the right place to do it.

Let me know what you think. I hope that this PR contributes some patterns that @jgoshi and @krwq can use for further refactoring.

### CODE REVIEW

File: build/Microsoft.DotNet.Cli.Version.props [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+      <Channel>master</Channel>
+      <BranchName>master</BranchName>
+
+      <CliVersionPrefix>$(VersionMajor).$(VersionMinor).$(VersionPatch)-$(ReleaseSuffix)</CliVersionPrefix>
+      <SimpleVersion>$(VersionMajor).$(VersionMinor).$(VersionPatch).$(CommitCount)</SimpleVersion>
+      <VersionSuffix>$(ReleaseSuffix)-$(CommitCount)</VersionSuffix>
+      <NugetVersion>$(VersionMajor).$(VersionMinor).$(VersionPatch)-$(VersionSuffix)</NugetVersion>
+      
+      <SdkVersion>$(NugetVersion)</SdkVersion>
+      <SdkNugetVersion>$(SdkVersion)</SdkNugetVersion>
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Version.props]: Is there a reason we have this property?  Why can't the places just use `$(NuGetVersion)`?  I can understand having a `NuGetVersion` and a `SdkVersion`, because our SdkVersion may not always be the NuGet version.  But having `SdkNuGetVersion` and `NuGetVersion` makes little sense to me.

File: build.proj [hunk 2]
```diff
@@ -1,10 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="BuildTheWholeCli" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
-  <Import Project="build/Microsoft.DotNet.Cli.InitRepo.props" />
-  <Import Condition=" Exists('$(GitCommitInfoProps)') Or '$(GeneratingPropsFile)' != 'true' " Project="$(GitCommitInfoProps)" />
-  <Import Condition=" Exists('$(HostInfoProps)') Or '$(GeneratingPropsFile)' != 'true' " Project="$(HostInfoProps)" />
-  <Import Project="build/Microsoft.DotNet.Cli.Monikers.props" />
```
Comment (eerhardt) [inline build.proj]: Can Monikers.props be deleted as well?  I don't see it removed.

File: build/Microsoft.DotNet.Cli.BranchInfo.props [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<Project>
+  <PropertyGroup>
+    <Channel>master</Channel>
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.BranchInfo.props]: This is duped with Version.props.

File: dir.props [hunk 4]
```diff
@@ -1,6 +1,22 @@
...
+  <Import Project="build/Microsoft.DotNet.Cli.BranchInfo.props" />
+
+  <Import Project="build/Microsoft.DotNet.Cli.DependencyVersions.props" />
+  <Import Project="build/Microsoft.DotNet.Cli.FileExtensions.props" />
+  <Import Project="build/Microsoft.DotNet.Cli.OutputDirectories.props" />
+  <Import Project="build/Microsoft.DotNet.Cli.Stage0.props" />
+  <Import Project="build/Microsoft.DotNet.Cli.Version.props" />
+  <Import Project="build/Microsoft.DotNet.Cli.Branding.props" />
+  
   <!-- NuGet Information -->
```
Comment (eerhardt) [inline dir.props]: This comment is pretty out of date.

File: build/Microsoft.DotNet.Cli.BranchInfo.props (hunk 3 above)
Comment (livarcocc) [inline build/Microsoft.DotNet.Cli.BranchInfo.props]: Shouold we remove BranchInfo.txt and just keep the props file? Where else is that txt being used?

File: build.proj (hunk 2 above)
Comment (livarcocc) [inline build.proj]: If Branding is the new monikers, which looks like it, I second deleting Monikers.props.

Down with duplication.

File: build/Microsoft.DotNet.Cli.Branding.props [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (livarcocc) [inline build/Microsoft.DotNet.Cli.Branding.props]: Do we need the xml?

File: build/Microsoft.DotNet.Cli.Version.props [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      <ReleaseSuffix>alpha</ReleaseSuffix>
+      <Channel>master</Channel>
+      <BranchName>master</BranchName>
+
+      <CliVersionPrefix>$(VersionMajor).$(VersionMinor).$(VersionPatch)-$(ReleaseSuffix)</CliVersionPrefix>
+      <SimpleVersion>$(VersionMajor).$(VersionMinor).$(VersionPatch).$(CommitCount)</SimpleVersion>
+      <VersionSuffix>$(ReleaseSuffix)-$(CommitCount)</VersionSuffix>
+      <NugetVersion>$(VersionMajor).$(VersionMinor).$(VersionPatch)-$(VersionSuffix)</NugetVersion>
+      
+      <SdkVersion>$(NugetVersion)</SdkVersion>
```
Comment (livarcocc) [inline build/Microsoft.DotNet.Cli.Version.props]: This one isn't needed either. They are all just dups of NuGetVersino. I ran into this on my DropSuffix change, but didn't feel like removing it as part of that change.

File: dir.props [hunk 7]
```diff
@@ -1,6 +1,22 @@
...
   <PropertyGroup>
     <RepoRoot>$(MSBuildThisFileDirectory)</RepoRoot>
```
Comment (livarcocc) [inline dir.props]: This is duplicated.

File: build.proj (hunk 2 above)
Comment (piotrpMSFT) [inline build.proj]: done.

File: build/Microsoft.DotNet.Cli.BranchInfo.props (hunk 3 above)
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.BranchInfo.props]: Where is BranchInfo.txt???

Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.BranchInfo.props]: > Where is BranchInfo.txt???

Deleted.  You're welcome.

**Comment by piotrpMSFT** [comment]
@eerhardt you've got it. Is there an interesting test or even `dotnet add` flavor that would be useful here? I bet @blackdwarf has given some thought to `automatically adding relevant content to sln files`

**Comment by piotrpMSFT** [comment]
@dotnet-bot test Debian8.2 x64 Debug Build please.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test Ubuntu x64 Release Build please.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test Windows_NT x86 Debug Build please.

**Comment by piotrpMSFT** [comment]
@mmitche we seem to be having networking issues in CI...

**Comment by piotrpMSFT** [comment]
@dotnet-bot test Windows_NT x86 Debug Build please.

File: test/dotnet.Tests/PackagedCommandTests.cs [hunk 8]
```diff
@@ -65,19 +65,19 @@ public void TestProjectToolIsAvailableThroughDriver(string appName, bool useCurr
             var testInstance = TestAssets.Get(appName)
                 .CreateInstance()
                 .WithSourceFiles()
-                .WithRestoreFiles();
+                .WithNuGetConfig(new RepoDirectoriesProvider().TestPackages);
```
Comment (eerhardt) [inline test/dotnet.Tests/PackagedCommandTests.cs]: I'm not sure why your original change (Prepare.targets) needs these changes. They seem unrelated.

Comment (piotrpMSFT) [inline test/dotnet.Tests/PackagedCommandTests.cs]: They are unrelated. It's very unclear to me how this restore worked in the first place, it follows a different pattern than other tests that share the intent of restoring within the test execution directory. Restoring prior to editing the project also seemed unnecessary and potentially side-effecting. I suspect this worked previously because the package already made it into my packages directory due to the initial restore, but that implicit dependency has shown itself to be fragile in other parts of the system [LZMA tool, CrossGen] and I'd prefer to fix the symptoms by cleaning up.

**Comment by mmitche** [comment]
@piotrpMSFT What was the error?

**Comment by piotrpMSFT** [comment]
@mmitche I saw a bunch of builds fail due to asset download errors. The asset URLs were correct and the assets were downloadable on local dev boxes. The issue seems to have resolved itself.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test RHEL7.2 x64 Release Build please.

**Comment by mmitche** [comment]
@piotrpMSFT Okay

There may be some network or other issues going on. I'm seeing greatly increased times to provision machines in some cases.

**Comment by jgoshi** [comment]
test Ubuntu16.04 x64 Debug Build
